### PR TITLE
Read dt in python. Closes #62

### LINKED ISSUE #62: Access dt from python

### PR DESCRIPTION
**Pull request by kjellkongsvik**

### CODE REVIEW

File: python/segyio/segy.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 class SegyFile(object):
 
-    def __init__(self, filename, mode, iline=189, xline=193, t0=1111.0):
+    def __init__(self, filename, mode, iline=189, xline=193):
```
Comment (jepebe) [inline python/segyio/segy.py]: I think that it should be possible to set `t0` unless it is trivial and guaranteed to get that value from the headers. The default should probably be `none` to allow to find in headers and default to `0`, while a value of `0` is specific.

File: python/segyio/segy.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def __init__(self, filename, mode, iline=189, xline=193, t0=1111.0):
         self._offsets = None
         self._ext_headers = None
         self._samples = None
+        self._dt = None
```
Comment (jepebe) [inline python/segyio/segy.py]: Should it be possible to override `dt`?

Comment (jokva) [inline python/segyio/segy.py]: Not with our current needs.

File: python/segyio/segy.py (hunk 1 above)
Comment (jokva) [inline python/segyio/segy.py]: Seems reasonable to me. I don't think it's a blocker though.

File: python/segyio/_segyio.c [hunk 3]
```diff
@@ -479,6 +479,25 @@ static PyObject *py_trace_bsize(PyObject *self, PyObject *args) {
...
+    PyArg_ParseTuple(args, "Od", &file_capsule, &dt);
+    segy_file *p_FILE = get_FILE_pointer_from_capsule(file_capsule);
+
+    if (PyErr_Occurred()) { return NULL; }
+
+    int error = segy_sample_interval(p_FILE, &dt);
+    if (error != 0) { return py_handle_segy_error(error, errno); }
+
+    PyObject *dict = PyDict_New();
+    PyDict_SetItemString(dict, "dt", Py_BuildValue("d", dt));
```
Comment (jokva) [inline python/segyio/_segyio.c]: Why does this return a dict over a plain int?

File: src/segyio/segy.c [hunk 4]
```diff
@@ -658,39 +658,36 @@ static int segy_sample_interval( segy_file* fp, double* dt) {
...
-    } else if (trace_header_dt == 0.0) {
-        fprintf(stderr, "Trace sampling rate in trace header set to 0.0. Will use SEGY header sampling rate of: %f\n", binary_header_dt);
-        *dt = binary_header_dt;
-    } else if (trace_header_dt != binary_header_dt) {
-        fprintf(stderr, "Trace sampling rate in SEGY header and trace header are not equal. Will use SEGY header sampling rate of: %f\n",
-                binary_header_dt);
-        *dt = binary_header_dt;
+    // microseconds: us
+    int _binary_header_dt_us;
+    int _trace_header_dt_us;
```
Comment (jokva) [inline src/segyio/segy.c]: These symbols (underscore prefix'd) are reserved for the language implementation.

File: src/segyio/segy.c [hunk 5]
```diff
@@ -658,39 +658,36 @@ static int segy_sample_interval( segy_file* fp, double* dt) {
...
+
+    segy_get_bfield(bin_header, BIN_Interval, &_binary_header_dt_us);
+    segy_get_field(trace_header, TRACE_SAMPLE_INTERVAL, &_trace_header_dt_us);
+
+    // milliseconds: ms
+    double binary_header_dt_ms = _binary_header_dt_us/1000.0;
+    double trace_header_dt_ms = _trace_header_dt_us/1000.0;
+
+    if (_binary_header_dt_us == 0 && _trace_header_dt_us == 0) {
+        fprintf(stderr, "Trace sampling rate in SEGY header and first trace header set to 0.0. Will not set.\n");
```
Comment (jokva) [inline src/segyio/segy.c]: We shouldn't fprintf from inside the library. It's up to the wrapper code to report errors in a fashion that suits them

File: src/segyio/segy.c [hunk 6]
```diff
@@ -658,39 +658,36 @@ static int segy_sample_interval( segy_file* fp, double* dt) {
...
+    } else {
+        *dt = trace_header_dt_ms;
     }
 
     return 0;
 
 }
 
-int segy_sample_indexes( segy_file* fp, double* buf, double t0, size_t count) {
+int segy_sample_indexes( segy_file* fp, double* buf, double t0, double dt, size_t count) {
```
Comment (jokva) [inline src/segyio/segy.c]: Does this have to be `double`? The rest of the library assumes float.

Comment (kjellkongsvik) [inline src/segyio/segy.c]: agreed, but was somewhat mixed usage of double and float. t0 and sample_indexes is double. Can of course change all that.

File: src/segyio/segy.c (hunk 5 above)
Comment (kjellkongsvik) [inline src/segyio/segy.c]: Agreed. Will silently use fallback dt

File: python/segyio/_segyio.c (hunk 3 above)
Comment (jokva) [inline python/segyio/_segyio.c]: What error objects? You're not returning one now. You can also use exceptions.

File: src/segyio/segy.c (hunk 5 above)
Comment (jokva) [inline src/segyio/segy.c]: Not necessarily silently (you can introduce a new error code, for instance).

File: python/segyio/_segyio.c (hunk 3 above)
Comment (kjellkongsvik) [inline python/segyio/_segyio.c]: if (error != 0) { return py_handle_segy_error(error, errno); }

Comment (jokva) [inline python/segyio/_segyio.c]: That will set an exception, not return an object. That function returns NULL.

File: src/segyio/segy.c (hunk 5 above)
Comment (kjellkongsvik) [inline src/segyio/segy.c]: That is possible. But probably not interesting.

**Comment by jepebe** [comment]
Looks good! Can a segy file, technically, have different *dt* for every trace?

**Comment by kjellkongsvik** [comment]
@jepebe yes, can have a different dt for every trace. We only check binary header and first trace.

**Pull request merged**